src/providers.rs

```diff
@@ -417,255 +313,18 @@ pub fn get_known_chain_id(service: &RpcService) -> Option<u64> {
-            p.borrow_mut().insert(provider_id, provider)
-        });
-        log!(
-            INFO,
-            "[{}] Withdrawing {} cycles from provider {} to canister: {}",
-            caller,
-            amount,
-            provider_id,
-            canister_id,
-        );
-        #[derive(CandidType)]
-        struct DepositCyclesArgs {
-            canister_id: Principal,
-        }
-        match ic_cdk::api::call::call_with_payment128(
-            Principal::management_canister(),
-            "deposit_cycles",
-            (DepositCyclesArgs { canister_id },),
-            amount,
-        )
-        .await
-        {
-            Ok(()) => add_metric!(cycles_withdrawn, amount),
-            Err(err) => {
-                // Refund on failure to send cycles
-                log!(
-                    INFO,
-                    "[{}] Unable to send {} cycles from provider {}: {:?}",
-                    canister_id,
-                    amount,
-                    provider_id,
-                    err
-                );
-                // Protect against re-entrancy
-                let provider = PROVIDERS.with(|p| {
-                    p.borrow()
-                        .get(&provider_id)
-                        .ok_or(ProviderError::ProviderNotFound)
-                });
-                let mut provider = provider.expect("Provider not found during refund, cycles lost");
-                PROVIDERS.with(|p| {
-                    provider.cycles_owed += amount;
-                    p.borrow_mut().insert(provider_id, provider)
-                });
-            }
-        };
-    } else {
-        ic_cdk::trap("You are not authorized: check provider owner");
-    }
-}
-
-pub fn set_service_provider(service: &RpcService, provider: &Provider) {
-    log!(
-        INFO,
-        "Updating service {:?} to use provider: {}",
-        service,
-        provider.provider_id
-    );
-    if let Some(chain_id) = get_known_chain_id(service) {
-        if chain_id != provider.chain_id {
-            ic_cdk::trap(&format!(
-                "Mismatch between service and provider chain ids ({} != {})",
-                chain_id, provider.chain_id
-            ))
-        }
-    }
-    SERVICE_PROVIDER_MAP.with(|mappings| {
-        mappings
-            .borrow_mut()
-            .insert(StorableRpcService::new(service), provider.provider_id);
-    });
-}
-
 pub fn resolve_rpc_service(service: RpcService) -> Result<ResolvedRpcService, ProviderError> {
     Ok(match service {
-        RpcService::Chain(id) => ResolvedRpcService::Provider(PROVIDERS.with(|providers| {
-            let providers = providers.borrow();
-            Ok(providers
-                .iter()
-                .find(|(_, p)| p.primary && p.chain_id == id)
-                .or_else(|| providers.iter().find(|(_, p)| p.chain_id == id))
+        RpcService::Chain(id) => ResolvedRpcService::Provider(
+            find_provider(|p| p.chain_id == id)
```

Comment:
The logic is equivalent because we never ended up changing the `primary` flag on any provider in the production canister. 

We could use this opportunity to remove the `RpcService::Chain()` variant given the ambiguity. Because this update will make a few other breaking changes, we could include this in the migration steps for canisters to prepare for the new release.